File: packages/compose/nockerl-design-components/src/main/kotlin/com/dizyx/nockerl/design/components/NockerlSegmented.kt
package com.dizyx.nockerl.design.components

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.dizyx.nockerl.design.tokens.LocalNockerlColors
import com.dizyx.nockerl.design.tokens.NockerlControlShape
import com.dizyx.nockerl.design.tokens.NockerlMotionDuration
import com.dizyx.nockerl.design.tokens.NockerlMotionEasing
import com.dizyx.nockerl.design.tokens.nockerlLitSurface
import com.dizyx.nockerl.design.tokens.nockerlRecessedSurface

/**
 * The standardized **segmented control**: the ratified **sliding cyan pill**
 * (r2, B3): a recessed `canvasAlt` track well ("fields sink") on the shared
 * control radius, with ONE solid-accent indicator that **slides** between
 * segments. The fill never tweens (design-laws §7). The pill's OFFSET animates
 * (an interpolatable transform) and the labels cross-fade color; the indicator
 * wears the shared top catch-light like every solid-accent surface (B1/B2).
 *
 * The active label is contrast-picked against the solid accent; inactive labels
 * are muted canvas ink. The indicator radius nests one step inside the track
 * (control 12 − inset 2 = 10).
 *
 * @param T the option type (typically an enum).
 * @param options the segments, in order.
 * @param selected the currently-active option.
 * @param onSelect invoked with the tapped option.
 * @param label maps an option to its display string.
 * @param modifier outer modifier (pass `Modifier.fillMaxWidth()` for a full-width control).
 * @param enabled when `false`, the whole control is dimmed and non-interactive.
 * @param segmentIcon optional per-segment leading glyph (rendered beside the label
 *   at the size tier's icon size); `null` (default) = labels only.
 * @param segmentEnabled per-segment enablement: a disabled segment dims and
 *   ignores taps while the rest stay live; default = all enabled.
 * @param size the density tier: [NockerlSegmentedSize.MD] (default) or the dense
 *   [NockerlSegmentedSize.SM] for tight settings rows.
 */
@Composable
fun <T> NockerlSegmented(
    options: List<T>,
    selected: T,
    onSelect: (T) -> Unit,
    label: (T) -> String,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    segmentIcon: ((T) -> ImageVector?)? = null,
    segmentEnabled: (T) -> Boolean = { true },
    size: NockerlSegmentedSize = NockerlSegmentedSize.MD,
) {
    val colors = LocalNockerlColors.current
    val alpha = if (enabled) 1f else DISABLED_ALPHA
    val trackColor = colors.canvasAlt.copy(alpha = colors.canvasAlt.alpha * alpha)
    val pillColor = colors.accentPrimary.copy(alpha = colors.accentPrimary.alpha * alpha)
    val selectedIndex = options.indexOf(selected).coerceAtLeast(0)
    val innerShape = nockerlRoundedShape(SEGMENT_INNER_RADIUS)

    Box(
        modifier =
            modifier
                .nockerlRecessedSurface(shape = NockerlControlShape)
                .background(trackColor)
                .padding(SEGMENT_TRACK_INSET),
    ) {
        BoxWithConstraints {
            // The sliding pill needs a BOUNDED width to compute equal segments:
            // callers pass `fillMaxWidth()` (or a set width); in a degenerate
            // unbounded host the indicator is skipped rather than crashing.
            val bounded = constraints.hasBoundedWidth
            val segmentWidth = if (bounded) maxWidth / options.size else 0.dp
            // The ONE pill: its position animates (interpolatable transform);
            // the solid fill itself never tweens (law §7).
            val indicatorOffset by animateDpAsState(
                targetValue = segmentWidth * selectedIndex,
                animationSpec = tween(NockerlMotionDuration.baseMs, easing = NockerlMotionEasing.standard),
                label = "nockerlSegmentedSlide",
            )

            Box(modifier = Modifier.height(IntrinsicSize.Min)) {
                if (bounded) {
                    Box(
                        modifier =
                            Modifier
                                .offset(x = indicatorOffset)
                                .width(segmentWidth)
                                .fillMaxHeight()
                                .clip(innerShape)
                                .background(pillColor)
                                // Same surfaceHighlight token as the web (the ruling:
                                // one identical catch-light, no per-platform drift).
                                .nockerlLitSurface(shape = innerShape),
                    )
                }
                Row {
                    options.forEach { option ->
                        val active = option == selected
                        val segmentLive = enabled && segmentEnabled(option)
                        // A disabled SEGMENT dims below the control's own alpha.
                        val segmentAlpha = if (segmentLive) alpha else alpha * DISABLED_ALPHA
                        // Labels CROSS-FADE color as the pill slides under them.
                        val labelColor by animateColorAsState(
                            targetValue =
                                if (active) {
                                    pickOnAccent(colors.accentPrimary)
                                } else {
                                    colors.onCanvasMuted
                                },
                            animationSpec = tween(NockerlMotionDuration.baseMs, easing = NockerlMotionEasing.standard),
                            label = "nockerlSegmentedLabel",
                        )
                        val ink = labelColor.copy(alpha = labelColor.alpha * segmentAlpha)
                        Row(
                            modifier =
                                Modifier
                                    .weight(1f)
                                    .clip(innerShape)
                                    .clickable(enabled = segmentLive) { onSelect(option) }
                                    .padding(vertical = size.verticalPadding),
                            horizontalArrangement = androidx.compose.foundation.layout.Arrangement.Center,
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            val icon = segmentIcon?.invoke(option)
                            if (icon != null) {
                                Icon(
                                    imageVector = icon,
                                    contentDescription = null, // the label carries the name
                                    tint = ink,
                                    modifier = Modifier.size(size.iconSize),
                                )
                                Spacer(modifier = Modifier.width(6.dp))
                            }
                            Text(
                                text = label(option),
                                style = size.textStyle(),
                                fontWeight = if (active) FontWeight.Medium else FontWeight.Normal,
                                color = ink,
                            )
                        }
                    }
                }
            }
        }
    }
}

/** Track inset: the pill nests 2dp inside the recessed well. */
private val SEGMENT_TRACK_INSET = 2.dp

/** Indicator corner radius, one step tighter than the 12dp track (12 − 2). */
private val SEGMENT_INNER_RADIUS = 10.dp

/** The segmented density tiers: default vs dense settings rows. */
enum class NockerlSegmentedSize(
    internal val verticalPadding: androidx.compose.ui.unit.Dp,
    internal val iconSize: androidx.compose.ui.unit.Dp,
) {
    /** Dense: tight settings rows (labelSmall, 4dp vertical rhythm). */
    SM(verticalPadding = 4.dp, iconSize = 14.dp),

    /** Default (labelMedium, the ratified 6dp rhythm). */
    MD(verticalPadding = 6.dp, iconSize = 16.dp),
    ;

    @Composable
    internal fun textStyle() =
        when (this) {
            SM -> MaterialTheme.typography.labelSmall
            MD -> MaterialTheme.typography.labelMedium
        }
}

File: Sources/NockerlDesign/Components/NockerlSegmented.swift
// NockerlSegmented: the standardized segmented control on Swift. REDUCE-FILLS
// (ratified: "state by OUTLINE, not fill"): the ratified solid sliding cyan pill
// is retired. The SELECTED segment now reads via a CYAN BORDER + CYAN INK, and the track
// is a FLAT NEUTRAL HAIRLINE container (zero fill) grouping the segments as connected peers.
//
//   - Track: a NEUTRAL `divider` hairline container on the 12pt control radius. ZERO fill,
//     no recessed inset-shade. It groups the segments as connected peers; strip it and the
//     control degrades to Tabs (the connected-peers track IS what makes it "segmented").
//   - Indicator: the SELECTED segment wears a thin `accentPrimary` BORDER (no fill) that
//     SLIDES between segments via `matchedGeometryEffect`, a position-only transform
//     (design-laws §7); nothing tweens its color.
//   - Labels: selected = `accentPrimary` (cyan-on-plain, because the prior on-accent pick
//     is moot with no fill to sit on) at medium weight; unselected = muted canvas ink.
//
// Capability union (matching Compose `NockerlSegmented`): per-segment leading icons,
// per-segment disable, and the SM density tier for tight settings rows. Tokens-pure;
// feedback honors the platform (design-laws §7).

import SwiftUI

/// The segmented density tiers: the default control vs the dense tier
/// for tight settings rows.
public enum NockerlSegmentedSize {
    /// Dense: tight settings rows (10pt label, 4pt vertical rhythm).
    case sm
    /// Default (12pt label, the ratified 6pt rhythm).
    case md

    var fontSize: CGFloat {
        switch self {
        case .sm: return NockerlFontSize.size10
        case .md: return NockerlFontSize.size12
        }
    }

    var verticalPadding: CGFloat {
        switch self {
        case .sm: return NockerlSpace.space1
        case .md: return NockerlSpace.space1 + 2
        }
    }

    var iconSize: CGFloat {
        switch self {
        case .sm: return 11
        case .md: return 12
        }
    }
}

/// The token-driven segmented control, a cyan-outlined selected segment on a neutral
/// hairline track (reduce-fills: state by outline, not fill).
public struct NockerlSegmented<T: Hashable>: View {
    private let options: [T]
    private let selected: T
    private let onSelect: (T) -> Void
    private let label: (T) -> String
    private let icon: ((T) -> String?)?
    private let segmentEnabled: (T) -> Bool
    private let size: NockerlSegmentedSize

    @Environment(\.colorScheme) private var colorScheme
    @Environment(\.isEnabled) private var isEnabled
    @Namespace private var pillNamespace

    /// Create a segmented control.
    /// - Parameters:
    ///   - options: the segments, in order.
    ///   - selected: the currently-active option.
    ///   - label: maps an option to its display string.
    ///   - icon: optional per-segment leading SF Symbol name; `nil` = labels only.
    ///   - segmentEnabled: per-segment enablement. A disabled segment dims and
    ///     ignores taps while the rest stay live (defaults to all enabled).
    ///   - size: the density tier (defaults to ``NockerlSegmentedSize/md``).
    ///   - onSelect: invoked with the tapped option.
    public init(
        options: [T],
        selected: T,
        label: @escaping (T) -> String,
        icon: ((T) -> String?)? = nil,
        segmentEnabled: @escaping (T) -> Bool = { _ in true },
        size: NockerlSegmentedSize = .md,
        onSelect: @escaping (T) -> Void
    ) {
        self.options = options
        self.selected = selected
        self.label = label
        self.icon = icon
        self.segmentEnabled = segmentEnabled
        self.size = size
        self.onSelect = onSelect
    }

    public var body: some View {
        let palette = NockerlPalette.resolve(colorScheme)
        let innerShape = RoundedRectangle(
            cornerRadius: NockerlRadius.control - trackInset,
            style: .continuous
        )
        let trackShape = RoundedRectangle(cornerRadius: NockerlRadius.control, style: .continuous)

        HStack(spacing: 0) {
            ForEach(options, id: \.self) { option in
                segment(option, palette: palette, innerShape: innerShape)
            }
        }
        // The selected border SLIDES between segments, one animated transform (law §7).
        .animation(.nockerlStandard(NockerlMotionDuration.base), value: selected)
        .padding(trackInset)
        // Reduce-fills (law §6): the track is a FLAT NEUTRAL HAIRLINE container (ZERO fill,
        // no recessed inset-shade), grouping the segments as connected peers (strip it and the
        // control degrades to Tabs). Selection reads via the cyan segment border + ink.
        .clipShape(trackShape)
        .overlay {
            trackShape.strokeBorder(palette.divider, lineWidth: NockerlSpace.spacePx)
        }
        .opacity(isEnabled ? 1 : disabledAlpha)
    }

    @ViewBuilder
    private func segment(
        _ option: T,
        palette: NockerlPalette,
        innerShape: RoundedRectangle
    ) -> some View {
        let active = option == selected
        let live = isEnabled && segmentEnabled(option)
        // Reduce-fills: no fill to sit on, so the SELECTED ink is CYAN-on-plain (the prior
        // on-accent pick is moot); unselected stays neutral muted canvas ink.
        let ink: Color = active ? palette.accentPrimary : palette.onCanvasMuted

        ZStack {
            if active {
                // Reduce-fills (law §6): the SELECTED segment reads via a CYAN BORDER (no
                // fill). matchedGeometryEffect slides the border between segments, a
                // position-only transform (law §7), never a fill tween.
                // "which peer is active" is a SELECTION state, so the edge takes
                // `widthSelection` @ `opacity.selection` like every other choice border.
                innerShape
                    .strokeBorder(
                        palette.accentPrimary.opacity(NockerlBorderOpacity.selection),
                        lineWidth: NockerlBorder.widthSelection
                    )
                    .matchedGeometryEffect(id: "nockerlSegmentedPill", in: pillNamespace)
            }

            HStack(spacing: NockerlSpace.space1) {
                if let symbol = icon?(option) {
                    Image(systemName: symbol)
                        .font(.system(size: size.iconSize, weight: .medium))
                }
                Text(label(option))
                    .font(.nockerl(size: size.fontSize, weight: active ? .medium : .regular))
            }
            .foregroundColor(ink)
            .padding(.vertical, size.verticalPadding)
            .opacity(live ? 1 : disabledAlpha)
        }
        .frame(maxWidth: .infinity)
        .contentShape(Rectangle())
        .onTapGesture {
            if live { onSelect(option) }
        }
        .accessibilityAddTraits(active ? [.isSelected] : [])
        .accessibilityLabel(label(option))
    }

    /// The selected border nests 2pt inside the hairline track (border radius 12 − 2 = 10),
    /// so the sliding cyan edge never collides with the container's hairline.
    private var trackInset: CGFloat { NockerlSpace.space05 }

    private var disabledAlpha: Double { 0.38 }
}
